ND resharding op

philei-tt opened the pull request:
### Ticket
Link to Github Issue https://github.com/tenstorrent/tt-metal/issues/22235

### Problem description
This is a first iteration of support for resharding of nd-sharded tensors

### What's changed
- Add `is_valid_for_legacy_reshard ` to decide if we can use kernels for legacy resharding. 
- Add `nd_reshard_multicore_generic_naive` to copy from input to resharded output page-by-page (very similar to kernels from [test_tensor_accessor_on_device.cpp](https://github.com/tenstorrent/tt-metal/blob/main/tests/ttnn/unit_tests/gtests/accessor/test_tensor_accessor_on_device.cpp)

### TODO (will be done in the following PRs)
- [ ] Rewrite the whole Resharding OP through newer [tt-nn op registration API](https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/adding_new_ttnn_operation.html)
- [ ] Make sure that each core read or writes local pages. This should improve perf a lot. 
- [ ] Add support for cases when resharding causes tile-size change (happens sometimes in row-major layout)
- [ ] Reenable DRAM resharding tests when TensorAccessor is fixed for dram sharding (https://github.com/tenstorrent/tt-metal/pull/24949)

### Checklist
- [X] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes -- https://github.com/tenstorrent/tt-metal/actions/runs/16199904225
- [x] New/Existing tests provide coverage for changes

philei-tt committed:
Add naive nd reshard

philei-tt committed:
Fix case with padded shards, and enable row-major layout

philei-tt committed:
Clean up nd-resharding test

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/data_movement/sharded/reshard/device/reshard_program_factory.cpp:
```diff
@@ -704,4 +705,114 @@ operation::ProgramWithCallbacks reshard_multi_core(const Tensor& input, Tensor&
     }
 }
 
+operation::ProgramWithCallbacks nd_reshard_multicore_generic_naive(const Tensor& input, Tensor& output) {
```
can this go to a different file?

philei-tt committed:
Merge branch 'main' into philei/nd-resharding-op

philei-tt committed:
Move nd resharding program in a separate file

philei-tt committed:
Clean up include order in reshard op

philei-tt committed:
Remove hard-coded grid size

philei-tt committed:
Move nd-sharding test to test_nd_reshard.py. Add more edge cases:
Different input/output grid, More dtypes, shard shape > input shape,
DRAM, etc.

philei-tt committed:
Use all worker cores for nd-reshard. Skip DRAM test for now

philei-tt committed:
Merge branch 'main' into philei/nd-resharding-op

philei-tt committed:
Fix compilation issues caused by merging main in branch
